MudMenu: Fix PositionAtCursor

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
Mudmenu as a Context Menu and in some cases when using PositionAtCursor was not properly setting the position due to the args events relating to the container it's in not necessarily where the popover needed to nest. Our javascript would then jumble that up more and make it near impossible to have consistent behavior. These changes standardize anytime the exact position is known such as args.PageX and args.PageY to not allow javascript to determine position (except flipping behavior). This is accomplished by a class `mud-popover-position-override`

Additionally added a DataGrid Context Menu example in the Docs section and a MudTable (I don't recommend this, too laggy) and DataGrid example in the UnitTests.Viewer section.

Resolves #3060 
Resolves #3854
Resolves #10073
Resolves: #8124

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Tons of visual tests in first comment

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

versile2 committed:
Update scss file to update positioning when mud-tab contains a badge

versile2 committed:
test page

versile2 committed:
Merge branch 'dev' of https://github.com/versile2/MudBlazor into dev

versile2 committed:
changes to mudmenu popover positioning

versile2 commented:
![PreFix](https://github.com/user-attachments/assets/4db63c60-c085-4962-a049-ac2b0e1a95fa)
![PostFix](https://github.com/user-attachments/assets/1abd2d84-f6d0-49ef-bf13-ec478d76bd07)

versile2 committed:
Added a Unit Test to verify new class is getting put on

versile2 commented:
I don't know how to make codecov happy.

ScarletKuro commented on src/MudBlazor/Components/Menu/MudMenu.razor.cs:
```diff
@@ -340,7 +344,10 @@ public Task OpenMenuAsync(EventArgs args, bool temporary = false)
-            _popoverStyle = $"margin-top: {args?.OffsetY.ToPx()}; margin-left: {args?.OffsetX.ToPx()};";
+            var staticPositionCSS = "mud-popover-position-override";
+            // if we apply this class we hard coded left/top, js will skip reassignments
+            PopoverClass = $"{PopoverClass ?? string.Empty} {staticPositionCSS}";
```
We are not really allowed to write to components parameters directly, it messes up Blazor, and sadly MSFT didn't add yet an analyzer for that to notify developers.
I see two solutions for that:
1. You create same logic as for `_popoverStyle`, but `_popoverClass`, and to this variable the `PopoverClass` and your logic, BUT you probably would need to add `OnInitialized` and do `_popoverClass = PopoverClass` because unlike the style, the class can be overridden by the user. Then you would also need to change 
https://github.com/MudBlazor/MudBlazor/blob/df29a110b68c589d8c18f748f51214d70ea1ea8b/src/MudBlazor/Components/Menu/MudMenu.razor#L53 to `_popoverClass`

2. You create similar the protected `ActivatorClassname` your own one, and using the CssBuilder you merge the `_popoverClass` and `PopoverClass` via the `AddClass` you could probably just add an private bool that will conditionally add `mud-popover-position-override` instead of `_popoverClass`. Then you also change
https://github.com/MudBlazor/MudBlazor/blob/df29a110b68c589d8c18f748f51214d70ea1ea8b/src/MudBlazor/Components/Menu/MudMenu.razor#L53 to the classname you created. I recommend doing this one.

Hope you get idea

versile2 replied:
I totally forgot about that. I will make those changes.

versile2 replied:
I updated changes. I am having difficulty with the Unit Test, it's in the commit but any pointers? It seems like OpenMenuAsync should activate the context menu (it does run through the code to activate). But doesn't seem to update my comp or mudMenu

`
        [Test]
        public async Task OpenMenuAsync_Should_Set_FixedPosition()
        {
            // Arrange
            var comp = Context.RenderComponent<DataGridContextMenuTest>();
            var mudMenuContext = comp.Instance.ContextMenu;
            mudMenuContext.Should().NotBeNull();
            
            // Act
            await Context.Renderer.Dispatcher.InvokeAsync(() => mudMenuContext.OpenMenuAsync(new MouseEventArgs()));

            // find mud-menu element
            var mudMenu = comp.Find("#" + mudMenuContext.FieldId);
            
            // Assert
            mudMenuContext.AnchorOrigin.Should().Be(Origin.TopLeft);
            mudMenu.ClassList.Should().Contain("mud-popover-position-override");

            // Style is still null here and can't access _popoverStyle due to private
            //mudMenu.Style.Should().Contain("top:"));
            //mudMenu.Style.Should().Contain("left:"));

            await Context.Renderer.Dispatcher.InvokeAsync(() => mudMenuContext.CloseMenuAsync());
        }
`

ScarletKuro replied:
I'm looking

ScarletKuro replied:
The `mudMenu.ClassLis` returns the classes for menu div, not the popover, and the `PopoverClass` is applied to the popover itself.
I did currently a workaround as it's 3am for me... but ideally we should find the popover (it's currently missing the mudpopoverprovider in test), find that element and the the `ClassList` from it. I will look at that tmr.

versile2 committed:
Changed Parameter set

Moved set to a private bool that applies static class when it's static position.

ScarletKuro committed:
Merge branch 'dev' into fix/mudmenupopoverposition

ScarletKuro committed:
Update

ScarletKuro committed:
Fix warnings

ScarletKuro committed:
Fix formatting

ScarletKuro committed:
Improve

ScarletKuro committed:
nits

ScarletKuro commented:
This also fixes: https://github.com/MudBlazor/MudBlazor/issues/8124

versile2 committed:
Moved hardcoded AnchorOrigin to markup
Removed trailing ; as it is a duplicate

versile2 committed:
Updated new Docs Example

versile2 committed:
added outerhtml check for top/left designation

versile2 committed:
Added example from #8124

versile2 commented:
> I fixed all the problems etc. The style check you possible could grab from the popover if you want. But as you can see the codecov doesn't complain anymore.
> 
> But I think using the DataGrid in the `OpenMenuAsync_Should_Set_FixedPosition()` aka `DataGridContextMenuTest` is too heavy. I would rather just use the code for testing from the #8124
> 
> upd: anyway, going bed now, its very late and I'm doing more and more typos in the text.
> 
> P.S. I also noticed it's writing in `AnchorOrigin`, but considering it's old code and was there before, I'm not going to bother for now. This is task for later.

Thanks for this, helped probably more than you know as I learn the code. Moved the AnchorOrigin to the markup instead of the parameter. Also there is a new Doc example with this for the DataGridContextMenuExample. Finally, Added a Viewer item for #8124

ScarletKuro committed:
Fix test

ScarletKuro committed:
Replace test with MenuPositionAtCursorTest

ScarletKuro commented on src/MudBlazor.UnitTests.Viewer/TestComponents/Tabs/TabsWithBadgesAndLongTabNamesTest.razor:
```diff
@@ -0,0 +1,11 @@
+﻿@namespace MudBlazor.UnitTests.TestComponents
+
+<MudTabs>
```
I feel like this example is from the other PR? Can this be removed?

ScarletKuro commented on src/MudBlazor/Styles/components/_tabs.scss:
```diff
@@ -204,9 +204,21 @@
-  margin-left: 8px;
-  margin-inline-start: 8px;
-  margin-inline-end: unset;
+  &.mud-badge-root {
```
are these changes from the https://github.com/MudBlazor/MudBlazor/pull/10099?

ScarletKuro commented:
I'm reverting the `TabsWithBadgesAndLongTabNamesTest.razor` and `_tabs.scss`, they are clearly from the other PR https://github.com/MudBlazor/MudBlazor/pull/10099

ScarletKuro committed:
Revert some files

ScarletKuro committed:
Cleanup

ScarletKuro commented on src/MudBlazor.Docs/Pages/Components/DataGrid/Examples/DataGridContextMenuExample.razor:
```diff
@@ -0,0 +1,67 @@
+
+@inject ISnackbar Snackbar
+
+<MudDataGrid @ref="_dataGrid" T="ClientInfo" Items="Clients" Filterable="true"
```
cc @danielchalmers @henon do we need this example in the docs, or it's too specific?

henon replied:
I think this is a good example, especially because it is not trivial to implement this.

ScarletKuro commented:
This looks very good to me now.

ScarletKuro commented:
I also think we don't really need `DataTableContextMenuActivatorTest.razor` as it's duplicates the `MenuPositionAtCursorTest.razor` just wraps other component, don't think we need to test different combination in this case because it's the same thing.

ScarletKuro committed:
nits

ScarletKuro commented:
Just few quick questions before it gets merged

https://github.com/MudBlazor/MudBlazor/blob/fd7132a5df156fc9b160a1c1e6b7a99c48024c27/src/MudBlazor/TScripts/mudPopover.js#L182
https://github.com/MudBlazor/MudBlazor/blob/fd7132a5df156fc9b160a1c1e6b7a99c48024c27/src/MudBlazor/TScripts/mudPopover.js#L326

Shouldn't it be just `classList` considering
https://github.com/MudBlazor/MudBlazor/blob/fd7132a5df156fc9b160a1c1e6b7a99c48024c27/src/MudBlazor/TScripts/mudPopover.js#L173
and that the rest code is using it?
Or there is some trick behind it?

versile2 committed:
removed redundancy

versile2 commented:
no trick, it was called fullname sometiems and shortname sometimes. I just standardized all to just classList

ScarletKuro committed:
Remove DataTableContextMenuActivatorTest as it essentially the same as MenuPositionAtCursorTest

ScarletKuro committed:
nit

ScarletKuro commented:
Thanks!

meenzen commented:
The context menu example is broken, the menu is being rendered to the page:

![image](https://github.com/user-attachments/assets/83cac69e-1ed4-4124-9ef0-ecbe2e14d0e7)

When you click it the wasm app crashes.

versile2 commented:
> The context menu example is broken, the menu is being rendered to the page:
> 
> ![image](https://private-user-images.githubusercontent.com/22305878/390416797-83cac69e-1ed4-4124-9ef0-ecbe2e14d0e7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3MTg2MjUsIm5iZiI6MTczMjcxODMyNSwicGF0aCI6Ii8yMjMwNTg3OC8zOTA0MTY3OTctODNjYWM2OWUtMWVkNC00MTI0LTllZjAtZWNiZTJlMTRkMGU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI3VDE0Mzg0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkZGU5MzkxMmExZTM2NTVlMjQ5ZWRhNWRlZGE3ZGY0MWMyZjY1ZTZjNWFmMjk0ZmU4NDlkY2EwYTY0MWQ5N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sLzBXRQ5HvV6zeT7ubcR-KYxEBZXLic-PuuFLZROX-k)
> 
> When you click it the wasm app crashes.

@ScarletKuro 
In WASM any line that does this or similar causes the app to crash (is not particular to the context menu) and they all work in Server Mode.
await httpClient.GetFromJsonAsync<List<Element>>("webapi/periodictable")

ScarletKuro commented:
> > The context menu example is broken, the menu is being rendered to the page:
> > ![image](https://private-user-images.githubusercontent.com/22305878/390416797-83cac69e-1ed4-4124-9ef0-ecbe2e14d0e7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3MTg2MjUsIm5iZiI6MTczMjcxODMyNSwicGF0aCI6Ii8yMjMwNTg3OC8zOTA0MTY3OTctODNjYWM2OWUtMWVkNC00MTI0LTllZjAtZWNiZTJlMTRkMGU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI3VDE0Mzg0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkZGU5MzkxMmExZTM2NTVlMjQ5ZWRhNWRlZGE3ZGY0MWMyZjY1ZTZjNWFmMjk0ZmU4NDlkY2EwYTY0MWQ5N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sLzBXRQ5HvV6zeT7ubcR-KYxEBZXLic-PuuFLZROX-k)
> > When you click it the wasm app crashes.
> 
> @ScarletKuro In WASM any line that does this or similar causes the app to crash (is not particular to the context menu) and they all work in Server Mode. await httpClient.GetFromJsonAsync<List>("webapi/periodictable")

Are we talking about `MudBlazor.Docs.Wasm`? That's expected since it's doing API call, but there is no server running with the controlled, so just ignore it.

meenzen commented:
That invisible menu shouldn't be rendered in the first place

versile2 commented:
> That invisible menu shouldn't be rendered in the first place

If I understand what you are saying, all the popovers work that way, inspect a page with a mudselect, mudautocomplete, tooltip, etc find the popover provider element and every popover will be listed there just not visible. The visible ones will have a data-ticks attribute > 0

ScarletKuro commented:
> If I understand what you are saying, all the popovers work that way, inspect a page with a mudselect, mudautocomplete, tooltip, etc find the popover provider element and every popover will be listed there just not visible. The visible ones will have a data-ticks attribute > 0

Are we talking about the Popover Provider? Unfortunately, on this screen, I can't tell if it is coming from the Popover or not. If it is, then that’s expected and unrelated to this PR, it will list all popovers inside it as invisible with empty content. When `Open=true`, they become visible with their content.  
https://github.com/MudBlazor/MudBlazor/blob/43fbb956b46650885a4b25230fece4a1c81ed253/src/MudBlazor/Components/Popover/MudPopoverProvider.razor#L5-L17

This was architected this way long before I joined the team.

versile2 commented:
looks like what's mentioned isn't from the provider after further research but also it's required. Popover needs a parent to place itself. There's nearly nothing in it, the actual menu is with the provider.
```
<div id="_contextMenu" class="mud-menu"><!--!--><!--!--><button type="button" class="mud-button-root mud-button mud-button-text mud-button-text-default mud-button-text-size-medium mud-ripple" _bl_649=""><span class="mud-button-label"></span></button><!--!--><div id="popover-d0144102-d780-4ada-9c36-f1aeeba99cee" class="mud-popover-cascading-value"></div><!--!-->

    <!--!--><div class="mud-overlay" style=""></div></div>
```

henon commented:
So to summarize the invisible element needs to be there, but clicking it shouldn't cause a crash, I guess.

ScarletKuro commented:
I was able to reproduce the issue with this example:  
https://github.com/MudBlazor/MudBlazor/blob/43fbb956b46650885a4b25230fece4a1c81ed253/src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridContextMenuTest.razor

There is indeed a hidden clickable element that crashes both server-side and wasm.
The stack trace:

```js
blazor.web.js:1 
 crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Object reference not set to an instance of an object.
System.NullReferenceException: Object reference not set to an instance of an object.
   at MudBlazorWebAppAuto.Client.Pages.Counter.<BuildRenderTree>b__0_14(RenderTreeBuilder __builder3)
   at Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder.AddContent(Int32 sequence, RenderFragment fragment)
   at MudBlazor.MudMenuItem.<BuildRenderTree>b__53_0(RenderTreeBuilder __builder2)
   at Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder.AddContent(Int32 sequence, RenderFragment fragment)
   at MudBlazor.MudListItem`1[[System.Object, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<BuildRenderTree>b__141_3(RenderTreeBuilder __builder3) in C:\Users\KuroPC\source\repos\MudBlazor\src\MudBlazor\Components\List\MudListItem.razor:line 51
   at Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder.AddContent(Int32 sequence, RenderFragment fragment)
   at MudBlazor.MudText.<BuildRenderTree>b__37_0(RenderTreeBuilder __builder2)
   at Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder.AddContent(Int32 sequence, RenderFragment fragment)
   at MudBlazor.MudElement.BuildRenderTree(RenderTreeBuilder builder) in C:\Users\KuroPC\source\repos\MudBlazor\src\MudBlazor\Components\Element\MudElement.cs:line 99
   at Microsoft.AspNetCore.Components.ComponentBase.<.ctor>b__7_0(RenderTreeBuilder builder)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment, Exception& renderFragmentException)
```

When clicking on it, the popover opens **BUT** this PR is not the cause of the issue. Setting `PositionAtCursor` to `false` or completely reverting this commit does not fix it. Therefore, I am removing the regression label from this PR.

ScarletKuro commented:
In fact, it's even reproducible on v7: https://try.mudblazor.com/snippet/wkQolPwMqOAZytWz

ScarletKuro commented:
Null exception is fixed in https://github.com/MudBlazor/MudBlazor/pull/10352
